src/shims/posix/linux/foreign_items.rs

```diff
@@ -192,3 +201,61 @@ fn getrandom<'tcx>(
     this.write_scalar(Scalar::from_machine_usize(len, this), dest)?;
     Ok(())
 }
+
+fn futex<'tcx>(
+    this: &mut MiriEvalContext<'_, 'tcx>,
+    args: &[OpTy<'tcx, Tag>],
+    dest: PlaceTy<'tcx, Tag>,
+) -> InterpResult<'tcx> {
+    if args.len() < 4 {
+        throw_ub_format!("incorrect number of arguments for futex syscall: got {}, expected at least 4", args.len());
+    }
+    let addr = this.read_scalar(args[1])?.check_init()?;
+    let op = this.read_scalar(args[2])?.to_i32()?;
+    let val = this.read_scalar(args[3])?.to_i32()?;
+
+    this.memory.check_ptr_access(addr, Size::from_bytes(4), Align::from_bytes(4).unwrap())?;
+
+    let addr = addr.assert_ptr();
```

Comment:
This doesn't work if the user passed in an integer just cast to a pointer.

In practice, does this address have to be dereferencable?

Reply:
The value stored at the address is compared with a given value. FUTEX_WAIT returns EAGAIN if they are different.

Reply:
All right. This should likely use `force_ptr` then, or just `deref_operand`.

What is the type of the pointer?

Reply:
> What is the type of the pointer?

So that's a bit annoying: The kernel interprets it as a `int *`, but users pass in different types (e.g. `*const AtomicI32` or `*mut u32`, or some other (repr(transparent) wrapper struct) etc.). FUTEX_WAIT only reads from the address, so it must at least be valid/allocated and readable.

FUTEX_WAKE does not read from the futex. Users might even call this one using `*const c_void` or `*const ()`. It is okay to call it on addresses that are 'gone', e.g. after deallocating/unmapping them. Parking_lot specifically makes use of that by not bothering to keep ownership of the futex around, only the pointer, in the part that will do the waking. Calling it on an already re-used address is also fine even though that might trigger the wrong wakeup, because futexes may have suprious wakeups.

The pointer type will be even more ~~annoying~~ flexible on Windows: WaitByAddress allows you to pass the size of the pointee as argument (1, 2, 4, or  8), so the pointer is interpreted as something from `*const i8` through `*const i64`.

Reply:
Okay, this will be interesting. You picked a tough first contribution to Miri. ;)

Let's not worry about Windows just yet...

Reply:
> You picked a tough first contribution to Miri. ;)

Well #1562 is labeled as `E-easy`. Tricked me ;)

Reply:
Fair.^^ I guess in the future I should check how complicated the APIs are that the shim needs to implement... I didn't know anything about futexes before.

Thanks a ton for working on this, it really helps when implementing such APIs to have someone that actually knows them. :)